temporalio/worker/workflow_instance.py

```diff
@@ -0,0 +1,1299 @@
+            return await query_fn(*args)
+        else:
+            query_fn = cast(Callable, query_defn.fn)
+            return query_fn(*args)
+
+
+class _WorkflowOutboundImpl(WorkflowOutboundInterceptor):
+    def __init__(self, instance: _WorkflowInstanceImpl) -> None:
+        # We are intentionally not calling the base class's __init__ here
+        self._instance = instance
+
+    def info(self) -> temporalio.workflow.Info:
+        return self._instance._info
+
+    def start_activity(
+        self, input: StartActivityInput
+    ) -> temporalio.workflow.ActivityHandle:
+        return self._instance._outbound_schedule_activity(input)
+
+    async def start_child_workflow(
+        self, input: StartChildWorkflowInput
+    ) -> temporalio.workflow.ChildWorkflowHandle:
+        return await self._instance._outbound_start_child_workflow(input)
+
+    def start_local_activity(
+        self, input: StartLocalActivityInput
+    ) -> temporalio.workflow.ActivityHandle:
+        return self._instance._outbound_schedule_activity(input)
+
+
+class _TimerHandle(asyncio.TimerHandle):
+    def __init__(
+        self,
+        seq: int,
+        when: float,
+        callback: Callable[..., Any],
+        args: Sequence[Any],
+        loop: asyncio.AbstractEventLoop,
+        context: Optional[contextvars.Context],
+    ) -> None:
+        super().__init__(when, callback, args, loop, context)
+        self._seq = seq
+
+    def _build_start_command(
+        self, delay: float
+    ) -> temporalio.bridge.proto.workflow_commands.WorkflowCommand:
+        command = temporalio.bridge.proto.workflow_commands.WorkflowCommand()
+        command.start_timer.seq = self._seq
+        command.start_timer.start_to_fire_timeout.FromNanoseconds(int(delay * 1e9))
+        return command
+
+    def _build_cancel_command(
+        self,
+    ) -> temporalio.bridge.proto.workflow_commands.WorkflowCommand:
+        return temporalio.bridge.proto.workflow_commands.WorkflowCommand(
+            cancel_timer=temporalio.bridge.proto.workflow_commands.CancelTimer(
+                seq=self._seq
+            ),
+        )
+
+
+class _ActivityHandle(temporalio.workflow.ActivityHandle[Any]):
+    def __init__(
+        self,
+        instance: _WorkflowInstanceImpl,
+        seq: int,
+        input: Union[StartActivityInput, StartLocalActivityInput],
+        result_fut: asyncio.Future,
+        fn: Awaitable[Any],
+    ) -> None:
+        # TODO(cretz): Customize name in 3.9+?
+        super().__init__(fn)
+        self._instance = instance
+        self._seq = seq
+        self._input = input
+        self._result_fut = result_fut
+
+    def _resolve_success(self, result: Any) -> None:
+        if not self._result_fut.done():
+            self._result_fut.set_result(result)
+
+    def _resolve_failure(self, err: Exception) -> None:
+        if not self._result_fut.done():
+            self._result_fut.set_exception(err)
```

Comment:
While it may not hold here (I haven't checked), technically in Python there are other ways tasks and futures can be resolved locally (e.g. cancelled because they timed out), and Python throws if you try to resolve an already-resolved future. So this is just for safety. I _think_ I can assume that only we can resolve _this_ future, but I'm not comfortable enforcing that assumption instead of just being defensive (here and everywhere I resolve futures).

Reply:
OK. At minimum, it probably makes sense to log a warning message.

Alternatively, make it an assert that only runs during tests. If it turns out to be a bad assumption, just get rid of it.

Reply:
I tend to agree with @Sushisource if someone tried to resolve our future it's on them and we should just let the exception propagate instead of silently ignoring this.

Reply:
I have updated all places where futures shouldn't be able to be resolved outside of our code to no longer check whether they are done before resolving